### PR TITLE
middleware for content type and accept headers

### LINKED ISSUE #14054: Check conten-type and return 415 if not supported by route
### Describe the bug

Noticed that Prysm BN does not check the content type of the request body and just tries to deserialize it, causing strange errors.

E.g. Error on Lodestar when trying to call `getAttesterDuties` with SSZ request body 
```
[vc-09-geth-prysm-lodestar] May-28 22:38:48.648[]                error: Failed to download attester duties epoch=0 - getAttesterDuties failed with status 400: Could not decode request body: invalid character '>' looking for beginning of value
[vc-09-geth-prysm-lodestar] Error: getAttesterDuties failed with status 400: Could not decode request body: invalid character '>' looking for beginning of value
```

As per spec, this route does not require to support SSZ request bodies and by default Lodestar will be using JSON but it would still be nice if Prysm could return a 415 error as this would allow to gradually support SSZ for more routes as the client can retry the request with JSON body and cache for this route that SSZ is not supported, only sending JSON in subsequent requests.

### Has this worked before in a previous version?

Likely not

### 🔬 Minimal Reproduction

I've been using Kurtosis to run the tests (with custom lodestar image)
```yaml
participants:
  - el_type: geth
    el_image: ethereum/client-go:stable
    cl_type: lodestar
    cl_image: nflaig/lodestar:ssz-api
    vc_type: prysm
    vc_image: gcr.io/prysmaticlabs/prysm/validator:latest
    count: 1
  - el_type: geth
    el_image: ethereum/client-go:stable
    cl_type: prysm
    cl_image: gcr.io/prysmaticlabs/prysm/beacon-chain:latest
    vc_type: lodestar
    vc_image: nflaig/lodestar:ssz-api
    count: 1
```

### Error

Error itself is not really relevant for this issue

### Platform(s)

Linux (x86)

### What version of Prysm are you running? (Which release)

`gcr.io/prysmaticlabs/prysm/beacon-chain:latest` (v5.0.3)

### Anything else relevant (validator index / public key)?

_No response_

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Adding middleware for handling content type and accept header checks. 
Only impacts the beacon apis for now https://ethereum.github.io/beacon-APIs

- [ ] future needs to handle keymanager apis and validator apis

**Which issues(s) does this PR fix?**

Fixes #14054

**Other notes for review**

### CODE REVIEW

File: api/server/middleware/middleware.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+				for _, serverAcceptedType := range serverAcceptedTypes {
+					if strings.HasPrefix(acceptType, serverAcceptedType) {
+						accepted = true
+						break
+					}
+				}
+				if accepted {
+					break
+				}
+			}
```
Comment (rkapka) [inline api/server/middleware/middleware.go:93]: We should follow https://datatracker.ietf.org/doc/html/rfc2616#section-14.1. I don't think we have to support all edge cases, but `*/*` and `type "/" "*"` would be nice to have.

File: beacon-chain/rpc/endpoints.go [hunk 2]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			methods: []string{http.MethodGet},
 		},
 		{
 			template: "/eth/v1/beacon/blocks",
 			name:     namespace + ".PublishBlock",
-			handler:  server.PublishBlock,
-			methods:  []string{http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
+				middleware.ContentTypeHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
+				middleware.AcceptHeaderHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:536]:
```suggestion
				middleware.AcceptHeaderHandler([]string{api.JsonMediaType}),
```

File: beacon-chain/rpc/endpoints.go [hunk 3]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			methods: []string{http.MethodPost},
 		},
 		{
 			template: "/eth/v1/beacon/blinded_blocks",
 			name:     namespace + ".PublishBlindedBlock",
-			handler:  server.PublishBlindedBlock,
-			methods:  []string{http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
+				middleware.ContentTypeHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
+				middleware.AcceptHeaderHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:546]:
```suggestion
				middleware.AcceptHeaderHandler([]string{api.JsonMediaType}),
```

File: beacon-chain/rpc/endpoints.go [hunk 4]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			methods: []string{http.MethodPost},
 		},
 		{
 			template: "/eth/v2/beacon/blocks",
 			name:     namespace + ".PublishBlockV2",
-			handler:  server.PublishBlockV2,
-			methods:  []string{http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
+				middleware.ContentTypeHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
+				middleware.AcceptHeaderHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:556]:
```suggestion
				middleware.AcceptHeaderHandler([]string{api.JsonMediaType}),
```

File: beacon-chain/rpc/endpoints.go [hunk 5]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			methods: []string{http.MethodPost},
 		},
 		{
 			template: "/eth/v2/beacon/blinded_blocks",
 			name:     namespace + ".PublishBlindedBlockV2",
-			handler:  server.PublishBlindedBlockV2,
-			methods:  []string{http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
+				middleware.ContentTypeHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
+				middleware.AcceptHeaderHandler([]string{api.JsonMediaType, api.OctetStreamMediaType}),
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:566]:
```suggestion
				middleware.AcceptHeaderHandler([]string{api.JsonMediaType}),
```

File: beacon-chain/rpc/endpoints.go [hunk 6]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			},
+			handler: server.GetAttesterSlashings,
+			methods: []string{http.MethodGet},
 		},
 		{
 			template: "/eth/v1/beacon/pool/attester_slashings",
 			name:     namespace + ".SubmitAttesterSlashing",
-			handler:  server.SubmitAttesterSlashing,
-			methods:  []string{http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:686]: Missing accept header

File: beacon-chain/rpc/endpoints.go [hunk 7]
```diff
@@ -365,188 +486,291 @@ func (s *Service) beaconEndpoints(
...
+			},
+			handler: server.GetFinalityCheckpoints,
+			methods: []string{http.MethodGet},
 		},
 		{
 			template: "/eth/v1/beacon/states/{state_id}/validators",
 			name:     namespace + ".GetValidators",
-			handler:  server.GetValidators,
-			methods:  []string{http.MethodGet, http.MethodPost},
+			middleware: []mux.MiddlewareFunc{
```
Comment (rkapka) [inline beacon-chain/rpc/endpoints.go:750]: Missing accept header

**Review by rkapka — CHANGES_REQUESTED** [review]

File: beacon-chain/rpc/endpoints.go (hunk 2 above)
Comment (james-prysm) [reply beacon-chain/rpc/endpoints.go:536]: oops thanks

File: beacon-chain/rpc/endpoints.go (hunk 7 above)
Comment (james-prysm) [reply beacon-chain/rpc/endpoints.go:750]: good catch

**Review by rkapka — APPROVED** [review]

**Pull request merged**